File: 2017-10-06-android/CleanArchitectureAndTDDDemo/core/src/test/java/com/example/interactors/createcat/CreateCatInteractor.java
package com.example.interactors.createcat;

/**
 * Created by MyndDev on 06/10/2017.
 */

class CreateCatInteractor {

    CatCreatedPresenter presenter;

    public CreateCatInteractor(CatCreatedPresenter presenter) {
        this.presenter = presenter;
    }

    public void execute() {
        CreateCatResponse response = new CreateCatResponse();
        response.name = "Kang Kang 'The Great'";

        presenter.present(response);
    }
}

File: 2018-01-PSITS-SultanKudarat/practice-2018-01-13/Decoupled_Final/CatsProject.Domain/Interactors/CreateCat/CreateCatInteractor.cs
﻿using Domain.Entities;
using Domain.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Domain.Interactors.CreateCat
{
    public class CreateCatInteractor
    {
        private ICatsRepository _respository;

        public CreateCatInteractor(ICatsRepository repository)
        {
            _respository = repository;
        }

        public CreateCatResponse Execute(string name, string title)
        {
            Guid newCatId = Guid.NewGuid();
            Cat newCat = new Cat(newCatId)
            {
                Name = name,
                Title = title
            };

            _respository.Add(newCat);
            Cat catFromDb = _respository.Get(newCatId);

            return MakeResponse(catFromDb);
        }

        private CreateCatResponse MakeResponse(Cat catFromDb)
        {
            string grandioseName;
            if (string.IsNullOrEmpty(catFromDb.Title))
                grandioseName = catFromDb.Name;
            else
                grandioseName = string.Format("{0}, \"{1}\"", catFromDb.Name, catFromDb.Title);

            CreateCatResponse response = new CreateCatResponse
            {
                grandioseName = grandioseName,
                isCatImportant = grandioseName.Length <= 10
            };

            return response;
        }
    }
}
